### ISSUE TITLE
Cast for input function monkey-patch doesn't fix things as expected type changes

### ISSUE THREAD

**Issue by solsword** [python/mypy#17519]
**Bug Report**

I'm writing code that monkey-patches the built-in `input` function, like so:

```python
import builtins

def f(x: str) -> str:
    print(x)
    return 'hi'

builtins.input = f

print(input('what'))  # prints 'what' and then 'hi'
```

MyPy reasonably gives me a type error on the assignment to input, saying:

```txt
t.py:7: error: Incompatible types in assignment (expression has type "Callable[[str], str]", variable has type "Callable[[object], str]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

Okay, fine so I change my code like this, exactly matching the reported variable type:

```python
import builtins

from typing import cast, Callable

def f(x: str) -> str:
    print(x)
    return 'hi'

builtins.input = cast(Callable[[object], str], f)

print(input('what'))  # prints 'what' and then 'hi'
```

I'm now casting to exactly what the error message said was expected. However, I *still* get a mypy error on the same line, and the expected type has changed:

```txt
t.py:9: error: Incompatible types in assignment (expression has type "Callable[[object], str]", variable has type "Callable[[DefaultArg(object)], str]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

Looking it up, it seems like DefaultArg is from mypy_extensions? That wasn't obvious, but I can get this to work by doing:

```python
import builtins

from typing import cast, Callable
from mypy_extensions import DefaultArg

def f(x: str) -> str:
    print(x)
    return 'hi'

builtins.input = cast(Callable[[DefaultArg(object)], str], f)

print(input('what'))  # prints 'what' and then 'hi'
```

**To Reproduce**

Run the code snippets provided above through mypy by saving them to a file and running `mypy filename`.

**Expected Behavior**

If I cast to the type that the error message says it needs, then the type error should go away.

If an "optional extensions" type is required to correctly type code dealing with a core built-in function, it should be mentioned somewhere in the python.org core documentation maybe? I first turned to searching docs.python.org to figure out what DefaultArg was so I could figure out where to import it from, but there are zero hits. Web searches without careful quoting are near useless, since modern search engines find all sorts of conversations about default arguments that have nothing to do with mypy. Thankfully I was clued in enough to search my mypy documentation pages and finally found it. I realize this may not be fixable, but maybe some kind of reminder could be included in the error message when DefaultArg is required?

The minimal correct fix here should be for the original error message to include DefaultArg in the expected type, since defining a function which has no default for its first argument and setting that as `input` will cause crashes for people who expect the usual behavior.

**Actual Behavior**

First type error message is just wrong about the expected/required type.

**Your Environment**

- Mypy version used: 1.8.0, so I upgraded to 1.10.1 (current latest) and have the exact same issue.
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): none specified
- Python version used: 3.11.8

**Comment by JelleZijlstra** [comment]
The issue here is that `builtins.input` has a default for its argument, and your assigned function does not, so the type is incompatible.

However, ideally mypy should express that without using the undocumented `DefaultArg` extension.

**Comment by solsword** [comment]
Yes, mypy should be catching the lack of default here. That mistake in my example code is intended. The issue is how Mypy reports that error. Mypy DOES avoid surfacing the `DefaultArg`  extension in its initial message, perhaps assuming that it's better to report just which type of argument is required to avoid confusion (or perhaps there's something else weird going on there, I'm not sure). In any case, the initial error message is wrong, because it doesn't report the correct/full expected type, hence when I cast to the type that it first said it wanted, I continue to get an error.

IMO there shouldn't be any cases in which I see "mypy says X is required" and my reaction is "okay, I know this is fine, let me cast to X" and then after adding the cast mypy now says that Y is required and the types still aren't compatible. If Y is really what's required, that should be displayed in the initial error message.
